File: src/core/wasm/CablePhysics.ts
/**
 * ============================================================================
 * CABLE PHYSICS ENGINE (TYPESCRIPT VERLET INTEGRATION ENGINE)
 * ============================================================================
 * Provides dual-mode particle Verlet integration for visual graph wires:
 * - Native JavaScript/TypeScript simulation for instant web interactivity
 * - Exact mathematical parity with C++ CablePhysics kernel
 * - Settling time < 300ms, distance constraint relaxation, and tunable profiles
 * ============================================================================
 */

import { Point2D, SplineResult, SplineSolver } from "./SplineSolver";

export type WireType =
  | "exec"
  | "string"
  | "number"
  | "boolean"
  | "object"
  | "array"
  | "default";

export interface CableParticle {
  position: Point2D;
  oldPosition: Point2D;
  acceleration: Point2D;
  isPinned: boolean;
}

export interface CableConfig {
  particleCount: number;
  gravity: number; // Downward gravity in px/s^2
  damping: number; // Velocity damping factor [0, 1]
  stiffness: number; // Constraint resolution strength [0, 1]
  constraintIterations: number;
  restLengthFactor: number;
}

export class CablePhysics {
  private config: CableConfig;
  private particles: CableParticle[] = [];
  private segmentLength: number = 10.0;
  private startPin: Point2D = { x: 0, y: 0 };
  private endPin: Point2D = { x: 0, y: 0 };

  constructor(type: WireType = "default", particleCount: number = 16) {
    this.config = CablePhysics.getConfigForWireType(type);
    if (particleCount > 2) {
      this.config.particleCount = particleCount;
    }
  }

  public static getConfigForWireType(type: WireType): CableConfig {
    switch (type) {
      case "exec":
        return {
          particleCount: 16,
          gravity: 90.0,
          damping: 0.22,
          stiffness: 0.95,
          constraintIterations: 5,
          restLengthFactor: 1.01,
        };
      case "boolean":
        return {
          particleCount: 16,
          gravity: 160.0,
          damping: 0.14,
          stiffness: 0.88,
          constraintIterations: 4,
          restLengthFactor: 1.04,
        };
      case "number":
      case "string":
      case "object":
      case "array":
      case "default":
      default:
        return {
          particleCount: 16,
          gravity: 200.0,
          damping: 0.12,
          stiffness: 0.82,
          constraintIterations: 4,
          restLengthFactor: 1.06,
        };
    }
  }

  public initialize(start: Point2D, end: Point2D, customConfig?: Partial<CableConfig>): void {
    if (customConfig) {
      this.config = { ...this.config, ...customConfig };
    }
    this.startPin = { ...start };
    this.endPin = { ...end };

    const count = Math.max(2, this.config.particleCount);
    this.particles = [];

    const dx = end.x - start.x;
    const dy = end.y - start.y;
    const chord = Math.sqrt(dx * dx + dy * dy);
    const totalLength = Math.max(chord, chord * this.config.restLengthFactor);
    this.segmentLength = totalLength / (count - 1);

    for (let i = 0; i < count; i++) {
      const t = i / (count - 1);
      const sag = 4.0 * t * (1.0 - t) * (this.config.gravity * 0.05);
      const px = start.x + dx * t;
      const py = start.y + dy * t + sag;
      const pinned = i === 0 || i === count - 1;

      this.particles.push({
        position: { x: px, y: py },
        oldPosition: { x: px, y: py },
        acceleration: { x: 0, y: 0 },
        isPinned: pinned,
      });
    }
  }

  public setEndpoints(start: Point2D, end: Point2D): void {
    const count = this.particles.length;
    if (count >= 2) {
      const dStartX = start.x - this.startPin.x;
      const dStartY = start.y - this.startPin.y;
      const dEndX = end.x - this.endPin.x;
      const dEndY = end.y - this.endPin.y;

      const dx = end.x - start.x;
      const dy = end.y - start.y;
      const chord = Math.sqrt(dx * dx + dy * dy);
      const totalLength = Math.max(chord, chord * this.config.restLengthFactor);
      this.segmentLength = totalLength / (count - 1);

      for (let i = 1; i < count - 1; i++) {
        const t = i / (count - 1);
        const shiftX = (1.0 - t) * dStartX + t * dEndX;
        const shiftY = (1.0 - t) * dStartY + t * dEndY;
        this.particles[i].position.x += shiftX;
        this.particles[i].position.y += shiftY;
        this.particles[i].oldPosition.x += shiftX;
        this.particles[i].oldPosition.y += shiftY;
      }

      this.particles[0].position = { ...start };
      this.particles[0].oldPosition = { ...start };
      this.particles[0].isPinned = true;

      const lastIdx = count - 1;
      this.particles[lastIdx].position = { ...end };
      this.particles[lastIdx].oldPosition = { ...end };
      this.particles[lastIdx].isPinned = true;
    }
    this.startPin = { ...start };
    this.endPin = { ...end };
  }

  public resetAlongCurve(spline: SplineResult): void {
    const count = this.particles.length;
    if (count < 2) return;

    for (let i = 0; i < count; i++) {
      const normDist = i / (count - 1);
      const t = SplineSolver.getTForNormalizedArcLength(spline.arcLengthTable, normDist);
      const pt = SplineSolver.evaluateBezier(spline.p0, spline.p1, spline.p2, spline.p3, t);

      this.particles[i].position = { ...pt };
      this.particles[i].oldPosition = { ...pt };
      this.particles[i].acceleration = { x: 0, y: 0 };
      this.particles[i].isPinned = i === 0 || i === count - 1;
    }
  }

  public step(dt: number): void {
    if (this.particles.length === 0 || dt <= 0) return;

    const clampedDt = Math.min(dt, 0.033);
    this.applyVerlet(clampedDt);

    for (let iter = 0; iter < this.config.constraintIterations; iter++) {
      this.satisfyConstraints();
    }
  }

  private applyVerlet(dt: number): void {
    const dtSq = dt * dt;
    const dampFactor = Math.max(0, 1.0 - this.config.damping);

    for (const p of this.particles) {
      if (p.isPinned) continue;

      const vx = (p.position.x - p.oldPosition.x) * dampFactor;
      const vy = (p.position.y - p.oldPosition.y) * dampFactor;

      p.oldPosition.x = p.position.x;
      p.oldPosition.y = p.position.y;

      p.position.x += vx + p.acceleration.x * dtSq;
      p.position.y += vy + (p.acceleration.y + this.config.gravity) * dtSq;

      p.acceleration.x = 0;
      p.acceleration.y = 0;
    }
  }

  private satisfyConstraints(): void {
    const count = this.particles.length;
    if (count < 2) return;

    this.particles[0].position.x = this.startPin.x;
    this.particles[0].position.y = this.startPin.y;
    this.particles[count - 1].position.x = this.endPin.x;
    this.particles[count - 1].position.y = this.endPin.y;

    for (let i = 0; i < count - 1; i++) {
      const p1 = this.particles[i];
      const p2 = this.particles[i + 1];

      const dx = p2.position.x - p1.position.x;
      const dy = p2.position.y - p1.position.y;
      const currentDist = Math.sqrt(dx * dx + dy * dy);

      if (currentDist < 0.0001) continue;

      const diff = (currentDist - this.segmentLength) / currentDist;
      const scalar = 0.5 * diff * this.config.stiffness;

      const offsetX = dx * scalar;
      const offsetY = dy * scalar;

      if (!p1.isPinned && !p2.isPinned) {
        p1.position.x += offsetX;
        p1.position.y += offsetY;
        p2.position.x -= offsetX;
        p2.position.y -= offsetY;
      } else if (p1.isPinned && !p2.isPinned) {
        p2.position.x -= offsetX * 2.0;
        p2.position.y -= offsetY * 2.0;
      } else if (!p1.isPinned && p2.isPinned) {
        p1.position.x += offsetX * 2.0;
        p1.position.y += offsetY * 2.0;
      }
    }
  }

  public isSettled(velocityThreshold: number = 0.5): boolean {
    for (const p of this.particles) {
      if (p.isPinned) continue;
      const vx = p.position.x - p.oldPosition.x;
      const vy = p.position.y - p.oldPosition.y;
      if (Math.sqrt(vx * vx + vy * vy) > velocityThreshold) {
        return false;
      }
    }
    return true;
  }

  public getKineticEnergy(): number {
    let total = 0;
    for (const p of this.particles) {
      const vx = p.position.x - p.oldPosition.x;
      const vy = p.position.y - p.oldPosition.y;
      total += vx * vx + vy * vy;
    }
    return total;
  }

  public getParticles(): readonly CableParticle[] {
    return this.particles;
  }

  public generateSvgPath(): string {
    const count = this.particles.length;
    if (count < 2) return "";

    let path = `M ${this.particles[0].position.x.toFixed(1)} ${this.particles[0].position.y.toFixed(1)}`;

    if (count === 2) {
      return `${path} L ${this.particles[1].position.x.toFixed(1)} ${this.particles[1].position.y.toFixed(1)}`;
    }

    for (let i = 0; i < count - 1; i++) {
      const p0 = i === 0 ? this.particles[i].position : this.particles[i - 1].position;
      const p1 = this.particles[i].position;
      const p2 = this.particles[i + 1].position;
      const p3 = i + 2 < count ? this.particles[i + 2].position : p2;

      const cp1x = (p1.x + (p2.x - p0.x) / 6.0).toFixed(1);
      const cp1y = (p1.y + (p2.y - p0.y) / 6.0).toFixed(1);
      const cp2x = (p2.x - (p3.x - p1.x) / 6.0).toFixed(1);
      const cp2y = (p2.y - (p3.y - p1.y) / 6.0).toFixed(1);
      const p2x = p2.x.toFixed(1);
      const p2y = p2.y.toFixed(1);

      path += ` C ${cp1x} ${cp1y}, ${cp2x} ${cp2y}, ${p2x} ${p2y}`;
    }

    return path;
  }
}

File: DOCS/Initial/decisions/0001-wasm-archive/src/CablePhysics.cpp
#include "CablePhysics.hpp"
#include <sstream>
#include <iomanip>

namespace WebAppEngine {

CablePhysics::CablePhysics(WireType type, int particleCount) {
    CableConfig config = getConfigForWireType(type);
    if (particleCount > 2) {
        config.particleCount = particleCount;
    }
    m_config = config;
}

CableConfig CablePhysics::getConfigForWireType(WireType type) {
    CableConfig config;
    switch (type) {
        case WireType::Exec:
            // Stiffer, fast recovery, reduced sag
            config.gravity = 90.0f;
            config.damping = 0.22f;
            config.stiffness = 0.95f;
            config.constraintIterations = 5;
            config.restLengthFactor = 1.01f;
            break;

        case WireType::DataBoolean:
            config.gravity = 160.0f;
            config.damping = 0.14f;
            config.stiffness = 0.88f;
            config.constraintIterations = 4;
            config.restLengthFactor = 1.04f;
            break;

        case WireType::DataNumber:
        case WireType::DataString:
        case WireType::DataObject:
        case WireType::DataArray:
        case WireType::Default:
        default:
            // Organic smooth cable drape
            config.gravity = 200.0f;
            config.damping = 0.12f;
            config.stiffness = 0.82f;
            config.constraintIterations = 4;
            config.restLengthFactor = 1.06f;
            break;
    }
    return config;
}

void CablePhysics::initialize(const Point2D& start, const Point2D& end, const CableConfig& config) {
    m_config = config;
    m_startPin = start;
    m_endPin = end;

    const int count = std::max(2, m_config.particleCount);
    m_particles.clear();
    m_particles.reserve(count);

    float dx = end.x - start.x;
    float dy = end.y - start.y;
    float chord = std::sqrt(dx * dx + dy * dy);
    float totalLength = std::max(chord, chord * m_config.restLengthFactor);
    m_segmentLength = totalLength / static_cast<float>(count - 1);

    for (int i = 0; i < count; ++i) {
        float t = static_cast<float>(i) / static_cast<float>(count - 1);
        // Initial catenary / parabola sagging offset
        float sag = 4.0f * t * (1.0f - t) * (m_config.gravity * 0.05f);
        float px = start.x + dx * t;
        float py = start.y + dy * t + sag;

        bool pinned = (i == 0 || i == count - 1);
        m_particles.emplace_back(px, py, pinned);
    }
}

void CablePhysics::setEndpoints(const Point2D& start, const Point2D& end) {
    const size_t count = m_particles.size();
    if (count >= 2) {
        float dStartX = start.x - m_startPin.x;
        float dStartY = start.y - m_startPin.y;
        float dEndX = end.x - m_endPin.x;
        float dEndY = end.y - m_endPin.y;

        float dx = end.x - start.x;
        float dy = end.y - start.y;
        float chord = std::sqrt(dx * dx + dy * dy);
        float totalLength = std::max(chord, chord * m_config.restLengthFactor);
        m_segmentLength = totalLength / static_cast<float>(count - 1);

        for (size_t i = 1; i < count - 1; ++i) {
            float t = static_cast<float>(i) / static_cast<float>(count - 1);
            float shiftX = (1.0f - t) * dStartX + t * dEndX;
            float shiftY = (1.0f - t) * dStartY + t * dEndY;
            m_particles[i].position.x += shiftX;
            m_particles[i].position.y += shiftY;
            m_particles[i].oldPosition.x += shiftX;
            m_particles[i].oldPosition.y += shiftY;
        }

        m_particles.front().position = start;
        m_particles.front().oldPosition = start;
        m_particles.front().isPinned = true;

        m_particles.back().position = end;
        m_particles.back().oldPosition = end;
        m_particles.back().isPinned = true;
    }
    m_startPin = start;
    m_endPin = end;
}

void CablePhysics::resetAlongCurve(const SplineResult& spline) {
    const size_t count = m_particles.size();
    if (count < 2) return;

    for (size_t i = 0; i < count; ++i) {
        float normDist = static_cast<float>(i) / static_cast<float>(count - 1);
        float t = spline.arcLengthTable.getTForNormalizedArcLength(normDist);
        Point2D pt = SplineSolver::evaluateBezier(spline.p0, spline.p1, spline.p2, spline.p3, t);

        m_particles[i].position = pt;
        m_particles[i].oldPosition = pt;
        m_particles[i].acceleration = Point2D(0.0f, 0.0f);
        m_particles[i].isPinned = (i == 0 || i == count - 1);
    }
}

void CablePhysics::step(float dt) {
    if (m_particles.empty() || dt <= 0.0f) return;

    // Clamp dt to avoid explosion on background tab freeze
    float clampedDt = std::min(dt, 0.033f);

    applyVerlet(clampedDt);

    for (int iter = 0; iter < m_config.constraintIterations; ++iter) {
        satisfyConstraints();
    }
}

void CablePhysics::applyVerlet(float dt) {
    const float dtSq = dt * dt;
    const float dampFactor = std::max(0.0f, 1.0f - m_config.damping);

    for (auto& p : m_particles) {
        if (p.isPinned) continue;

        // Velocity = position - oldPosition
        float vx = (p.position.x - p.oldPosition.x) * dampFactor;
        float vy = (p.position.y - p.oldPosition.y) * dampFactor;

        p.oldPosition = p.position;

        // Verlet integration: x_new = x + v + a * dt^2
        p.position.x += vx + p.acceleration.x * dtSq;
        p.position.y += vy + (p.acceleration.y + m_config.gravity) * dtSq;

        // Reset per-step acceleration
        p.acceleration = Point2D(0.0f, 0.0f);
    }
}

void CablePhysics::satisfyConstraints() {
    const size_t count = m_particles.size();
    if (count < 2) return;

    // Pin endpoints first
    m_particles.front().position = m_startPin;
    m_particles.back().position = m_endPin;

    // Distance constraints
    for (size_t i = 0; i < count - 1; ++i) {
        auto& p1 = m_particles[i];
        auto& p2 = m_particles[i + 1];

        float dx = p2.position.x - p1.position.x;
        float dy = p2.position.y - p1.position.y;
        float currentDist = std::sqrt(dx * dx + dy * dy);

        if (currentDist < 0.0001f) continue;

        float diff = (currentDist - m_segmentLength) / currentDist;
        float scalar = 0.5f * diff * m_config.stiffness;

        float offsetX = dx * scalar;
        float offsetY = dy * scalar;

        if (!p1.isPinned && !p2.isPinned) {
            p1.position.x += offsetX;
            p1.position.y += offsetY;
            p2.position.x -= offsetX;
            p2.position.y -= offsetY;
        } else if (p1.isPinned && !p2.isPinned) {
            p2.position.x -= offsetX * 2.0f;
            p2.position.y -= offsetY * 2.0f;
        } else if (!p1.isPinned && p2.isPinned) {
            p1.position.x += offsetX * 2.0f;
            p1.position.y += offsetY * 2.0f;
        }
    }
}

bool CablePhysics::isSettled(float velocityThreshold) const {
    for (const auto& p : m_particles) {
        if (p.isPinned) continue;
        float vx = p.position.x - p.oldPosition.x;
        float vy = p.position.y - p.oldPosition.y;
        if (std::sqrt(vx * vx + vy * vy) > velocityThreshold) {
            return false;
        }
    }
    return true;
}

float CablePhysics::getKineticEnergy() const {
    float total = 0.0f;
    for (const auto& p : m_particles) {
        float vx = p.position.x - p.oldPosition.x;
        float vy = p.position.y - p.oldPosition.y;
        total += (vx * vx + vy * vy);
    }
    return total;
}

std::string CablePhysics::generateSvgPath() const {
    const size_t count = m_particles.size();
    if (count < 2) return "";

    std::ostringstream ss;
    ss << std::fixed << std::setprecision(1);

    ss << "M " << m_particles[0].position.x << " " << m_particles[0].position.y;

    if (count == 2) {
        ss << " L " << m_particles[1].position.x << " " << m_particles[1].position.y;
        return ss.str();
    }

    // Catmull-Rom to Cubic Bezier conversion for continuous smooth visual cable
    for (size_t i = 0; i < count - 1; ++i) {
        Point2D p0 = (i == 0) ? m_particles[i].position : m_particles[i - 1].position;
        Point2D p1 = m_particles[i].position;
        Point2D p2 = m_particles[i + 1].position;
        Point2D p3 = (i + 2 < count) ? m_particles[i + 2].position : p2;

        Point2D cp1(
            p1.x + (p2.x - p0.x) / 6.0f,
            p1.y + (p2.y - p0.y) / 6.0f
        );
        Point2D cp2(
            p2.x - (p3.x - p1.x) / 6.0f,
            p2.y - (p3.y - p1.y) / 6.0f
        );

        ss << " C " << cp1.x << " " << cp1.y << ", "
           << cp2.x << " " << cp2.y << ", "
           << p2.x << " " << p2.y;
    }

    return ss.str();
}

} // namespace WebAppEngine
